lib/AST/ASTImporter.cpp

```diff
@@ -4206,36 +4204,31 @@ Decl *ASTNodeImporter::VisitClassTemplateDecl(ClassTemplateDecl *D) {
 
   CXXRecordDecl *FromTemplated = D->getTemplatedDecl();
 
-  // Create the declaration that is being templated.
-  CXXRecordDecl *ToTemplated = cast_or_null<CXXRecordDecl>(
-        Importer.Import(FromTemplated));
-  if (!ToTemplated)
-    return nullptr;
-
-  // Resolve possible cyclic import.
-  if (Decl *AlreadyImported = Importer.GetAlreadyImportedOrNull(D))
-    return AlreadyImported;
-
   // Create the class template declaration itself.
   TemplateParameterList *TemplateParams =
       ImportTemplateParameterList(D->getTemplateParameters());
   if (!TemplateParams)
     return nullptr;
 
+  // Create the declaration that is being templated.
+  CXXRecordDecl *ToTemplated =
+      cast_or_null<CXXRecordDecl>(Importer.Import(FromTemplated));
+  if (!ToTemplated)
+    return nullptr;
+  Importer.Imported(FromTemplated, ToTemplated);
```

Comment:
Wouldn't VisitRecordDecl mark this imported? Isn't this a redundant call?

Reply:
Actually, this is redundant. But having redundant `Imported` calls is not a problem until the `To` is always the same.
Note w,e have an other redundancy in `Importer::Imprort(Decl*)` where we set the imported. In most cases this is redundant too, but better to have redundancy than missing just one case.

Reply:
Redundancy may not be bad, only here the call is immediately after the `Import` call and we (should) know that `Import` calls `Imported` always. (Otherwise there are a lot of places where `Import` is called and `Imported` is not called after it, somebody may think that here it is some special case.)

Reply:
Now I like it even less. Instead of removing a single `Imported` call it was inserted everywhere. One could ask: "Why is this all needed instead of doing it automatically in `Import`? What if the call to `Imported` is forgotten?" Alternative way of doing this: Do not call `Imported` but use an assertion to check if it was really imported.

Reply:
Regarding to your first comment: We have to ensure that after a node is created with `::Create` we immediately add that to the `ImportedDecls`, or at least before we would recursively descend to other imports.

If we forget the call to `Imported` then we cannot simply track which decl is the pair of the already visited and possibly imported "from" decl in the "to" context.

Reply:
Yes the `Imported` needs to be called after the `Create` calls, and with no other `Import` in between. But if we import something with `Import` the `Imported` call is not required afterwards, I see it as redundant code. There can be special cases when `Imported` is called with an already existing node (multiple nodes from the From-context are mapped to a single node in the To-context).